Protocol.md

```diff
@@ -279,6 +279,49 @@ The recipient:
 
   These script pub key forms include only standard forms accepted by the wider set of deployed Bitcoin clients in the network, which increase the chances of successful propagation to miners.
 
+## Optional features
+
+Optional features consist of "features" which parties can opt-in to using. These messages are not needed to facilitate construction of a DLCs.
+
+### The `close_dlc` Message
+
+This message contains information about mutual close, which allows the counterparty
+to broadcast a mutual closing transaction.
+
+1. type: ? (`close_dlc_v0`)
+2. data:
+   * [`32*byte`:`contract_id`]
+   * [`signature`:`close_signature`]
+   * [`u64`:`offerPayoutSatoshis`]
+   * [`u64`:`acceptPayoutSatoshis`]
+   * [`u64`:`fund_input_serial_id`]
+   * [`u16`:`num_funding_inputs`]
+   * [`num_funding_inputs*funding_input`:`funding_inputs`]
+
+
+`payout_spk` and `payout_serial_id` from `offer_dlc` as well as `payout_spk` and `payout_serial_id` from `accept_dlc` should be used for constructing the close transaction
+
+`fund_input_serial_id` is a randomly chosen number which uniquely identifies the funding output to be spent.
+Inputs in the closing transaction will be sorted by `fund_input_serial_id` and `input_serial_id` in `funding_inputs`.
+
+`funding_inputs` are extra inputs to mutual close to enable the sending party to nullify the transaction by double spending the inputs.
```

Comment:
I'm not sure you can really guarantee the double-spend without further checks on the payouts outputs `witnessScripts`. I think your counterparty might start to spend its payouts, even honestly, as a start of a new DLC or LN, thus falsifying your fee-estimation to successfully double-spend.

One mitigation could be to require to reveal `witnssScripts` and enforce there is a 1 OP_CSV to ensure children can be accepted as final in the network mempools.

Reply:
I'm not sure what is your concern, but the point of being able to double spend is in case your counter party doesn't broadcast the closing TX, so I feel if he would be spending the closing TX output it means that it's already broadcast and thus you don't need to worry about double spending?

Reply:
Agree with @Tibo-lg. Curious to know if there's a case that's not accounted for here. 

Generally the initiator should double-spend the `funding_inputs` in the case that the closing tx is not created in a timely manner (very dependent on the type of contract of course, and market conditions). This should be done shortly after sending the dlc close message, in the case that the counterparty doesn't mutually close the dlc. 

In the case that the counterparty does mutually close the dlc in a timely manner, then nullifying the transaction would no longer be a concern.

Reply:
Just to be clear, what do you qualify by "extra inputs", the funding outputs of a DLC funding transaction or inputs beyond the scope of the protocol ?

Reply:
These extra inputs are beyond the scope of the protocol. Just utxos that exist in a users wallet

Reply:
Thanks I see, though I think my point still hold.

Let's say you have a mutual closing transaction between Alice-Bob composed of 2-inputs and 2-outputs.

First input is the outpoint of the funding transaction output (the 2-of-2 P2WSH)
Second input is an input from Alice wallet.
First output is Alice's payout.
Second output is Bob's payout.

The mutual close is broadcasted by Alice with an absolute fee of 1000 sats.
Bob spends his own payout output with a child transaction paying an absolute fee of 1000 sats.
Alice attempts to double-spend the mutual close by spending her wallet input with a transaction paying an absolute fee of 1500 sats.

According BIP125, rule 3, a replacement transactions must pay an absolute fee higher than the replaced transactions. Alice attempts should fail as the fee is lower than the mutual close + Bob's child one ?

I think if we want to guarantee the replacement of the mutual close with an unilateral CET, I think we should ensure that
1) the payout outputs are encumbered by a 1 CSV (to expel any child from the network mempools)
2) the absolute fee, the feerate and replacement fee penalty of any CET transaction is higher than the mutual close one

However, I think fulfilling the second requirement is in contradiction with the goal of this proposal, namely fast-close in period of fees higher than the feerate with which CET have been counter-signed.

Let me know if you think the description is correct.

So I think a better behavior, instead of double-spending the mutual close with wallet inputs could be to CPFP the mutual close ? And if it doesn't work, assumes the CET has been broadcast across network mempools and CPFP the CET ?

Or we can swallow the bullet for now, ship it as it is, mark it as an issue in #132 and wait for package-relay simplifying that kind of issue :) ?

Reply:
I think I see the confusion. The initiator has a wallet input, but the reciprocator (the one broadcasting) does not. 

> So I think a better behavior, instead of double-spending the mutual close with wallet inputs could be to CPFP the mutual close ? And if it doesn't work, assumes the CET has been broadcast across network mempools and CPFP the CET ?

Additionally, the purpose of the wallet input is to nullify the mutual close in the case that the mutual close has not been broadcast at all. So it would not be possible to `CPFP the mutual close` if it has not been broadcast in the first place. 

> The mutual close is broadcasted by Alice with an absolute fee of 1000 sats.
> Bob spends his own payout output with a child transaction paying an absolute fee of 1000 sats.
> Alice attempts to double-spend the mutual close by spending her wallet input with a transaction paying an absolute fee of 1500 sats.

In the DLC Close spec proposed, only one party has a wallet input, and that would be the initiator. So if Alice is broadcasting the closing tx, she would be the reciprocator, not the initiator, and she would not have her own wallet input as part of the closing tx. Bob however would have a wallet input. 

So a feasible scenario similar to what you've proposed might look like this (assuming Alice is the initiator and Bob is the reciprocator and Alice is spending her wallet input because Bob did not broadcast mutual close in a timely manner):

First input is the outpoint of the funding transaction output (the 2-of-2 P2WSH)
Second input is an input from Alice wallet.
First output is Alice's payout.
Second output is Bob's payout.

Alice spends her wallet input with a transaction paying an absolute fee of 1500 sats.
Bob attempts to double-spend Alice's wallet input by broadcasting the mutual close with an absolute fee of 1000 sats.
Bob continues the attempt to double spend Alice's wallet input by spending his own payout output with a child transaction paying an absolute fee of 1000 sats.

In this instance, Alice would be spending her wallet input since Bob did not broadcast in a timely manner. 

What's not clear to me is whether BIP 125 rule 3 would make Bob's double-spend using the mutual close tx invalid since the absolute fees would be lower than Alice's original spend of her wallet input.

But regardless of whether Bob's broadcast of the mutual close is successful, Alice has either successfully spent her wallet input, nullifying the mutual close, or has forced Bob's hand to actual broadcast the mutual close in a timely manner, which is why Alice was spending her wallet input in the first place. 

In a real-world scenario, one could imagine a futures contract that Alice and Bob have entered into for a DLC. After some time inside the contract, Alice wishes to exit the position with Bob, and waits only 1 minute after passing a DlcClose message to him. After 1 minute, Alice does not detect the mutual close tx in the mempool, so Alice spends her wallet input and either she succeeds in nullifying the mutual close tx, or Bob successfully double-spends her wallet input with the mutual close tx which achieves Alice's goal in the first place.

Reply:
Thanks, now I think we have the same model in mind.

> What's not clear to me is whether BIP 125 rule 3 would make Bob's double-spend using the mutual close tx invalid since the absolute fees would be lower than Alice's original spend of her wallet input.

Currently yes, BIP 125 rule 3 would prevent Bob's double spend Alice input. Though if package relay is deployed Bob would be able to broadcast a mutual_close + CPFP and if mempool fee spikes at the same time, blocks Alice wallet input spending. Of course, Alice could override assuming she is able to observe Bob mutual_close + CPFP broadcast (an attacker could prevent this by broadcasting low-fees M1 inside Alice's mempool and pinning M2 across the rest of the network).

I think that case is good to know and to monitor when package relay is deployed.

> But regardless of whether Bob's broadcast of the mutual close is successful, Alice has either successfully spent her wallet input, nullifying the mutual close, or has forced Bob's hand to actual broadcast the mutual close in a timely manner, which is why Alice was spending her wallet input in the first place.

I think that's correct and it would be a good rational to add to the PR. Further, I think it could be good to specify an implementation-defined `nullifying_point` for Alice to enforce, similar in spirit to the `security_point` from `Non-Interactive-Protocol.md`. Eg, value could be dynamic in function of mempool fees changes nullifying the economic equilibrium of the DLC.

> In a real-world scenario, one could imagine a futures contract that Alice and Bob have entered into for a DLC. After some time inside the contract, Alice wishes to exit the position with Bob, and waits only 1 minute after passing a DlcClose message to him. After 1 minute, Alice does not detect the mutual close tx in the mempool, so Alice spends her wallet input and either she succeeds in nullifying the mutual close tx, or Bob successfully double-spends her wallet input with the mutual close tx which achieves Alice's goal in the first place.

Really agree on the long-term design that we should have a bunch of options, even negotiated with fee premiums to update or exit the DLC in function of other changes, such as mempool fee spikes or lightning routing fees ones :)

Reply:
Glad we're on the same page :) 

> Currently yes, BIP 125 rule 3 would prevent Bob's double spend Alice input. Though if package relay is deployed Bob would be able to broadcast a mutual_close + CPFP and if mempool fee spikes at the same time, blocks Alice wallet input spending. Of course, Alice could override assuming she is able to observe Bob mutual_close + CPFP broadcast (an attacker could prevent this by broadcasting low-fees M1 inside Alice's mempool and pinning M2 across the rest of the network).

Got it, so essentially as it stands, BIP 125 would favor Alice, and package relay would favor Bob. Either way, she achieves her goal.

> I think that's correct and it would be a good rational to add to the PR. Further, I think it could be good to specify an implementation-defined nullifying_point for Alice to enforce, similar in spirit to the security_point from Non-Interactive-Protocol.md. Eg, value could be dynamic in function of mempool fees changes nullifying the economic equilibrium of the DLC.

Great point, I've added `nullifying_point` and rationale to [`Non-Interactive-Protocol.md` ](https://github.com/discreetlogcontracts/dlcspecs/pull/170/files#diff-fb2caf21cc01d287e559f7cbfa10bae18416bc18449507e4849720f584e2655dR260) explaining the reasoning for spending funding-inputs. 

> Really agree on the long-term design that we should have a bunch of options, even negotiated with fee premiums to update or exit the DLC in function of other changes, such as mempool fee spikes or lightning routing fees ones :)

💯  I'm sure there's lots more that can be done and improved with future iterations of closing messages!